### PR TITLE
feat: add ADR for certs management

### PR DESCRIPTION
**Pull request by machi1990**
The ADR intends to supersede https://architecture.appservices.tech/adr/5/

### CODE REVIEW

**Comment by machi1990** [comment]
@emmanuelbernard @tombentley can I get a review on this? It is related to https://issues.redhat.com/browse/MGDSTRM-10159

**Comment by machi1990** [comment]
@pb82 @shawkins can you have a look?

Once accepted by you, i'll add another commit to mark the https://architecture.appservices.tech/adr/5/ as superseded by this one and I'll then add the status of the ADR as accepted.

File: _adr/89/index.adoc
Comment (pawelpaszki) [inline _adr/89/index.adoc:46]:
```suggestion
The proposed solution is going to use https://Letsencrypt.org[Letsencrypt] to automatically manage certificates.
```

File: _adr/89/index.adoc
Comment (pawelpaszki) [inline _adr/89/index.adoc:51]:
```suggestion
2. The fleet manager in its reconcilers will use https://github.com/caddyserver/certmagic[certmagic library] to generate certificate for each data plane and store them in a secure Storage. 
```

File: _adr/89/index.adoc
Comment (pawelpaszki) [inline _adr/89/index.adoc:59]:
```suggestion
4. Certificate renewals will also be handled automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 
```

File: _adr/89/index.adoc
Comment (pawelpaszki) [inline _adr/89/index.adoc:93]:
```suggestion
The Control plane then handles the pushing of the Certificate request CR, watching for the created secrets and storing them securely. 
```

**Comment by pawelpaszki** [comment]
@machi1990 - the ADR makes sense to me. I will let others review as well

**Comment by machi1990** [comment]
Thanks for the review @pawelpaszki and the suggestions. I'll apply them and wait for review from others as well.

File: _adr/89/index.adoc
Comment (emmanuelbernard) [inline _adr/89/index.adoc:89]: I understand but then let's put in the threat model that this  wildcard certificate of a lower level will allow impersonation of the enterprise kafka cluster.

File: _adr/89/index.adoc
Comment (machi1990) [reply _adr/89/index.adoc:52]: No, today all the routes for Kafka are a subdomain of `kafka.bf2.dev`. It'll only be the case going forward with this ADR is implemented. This is one of the reason why performing a transparent migration is hard.

File: _adr/89/index.adoc
Comment (machi1990) [reply _adr/89/index.adoc:25]: It is briefly stated in the goals. I can additionally copy & paste the problem as state in the ADR 5, clearly stating that the issue at the moment is that the process is manual and it exposes shared the certs to customer data planes. What do you think?

File: _adr/89/index.adoc
Comment (machi1990) [reply _adr/89/index.adoc:59]: +1 makes sense. 

in the long run, we hope there will be a central service for ingress management e.g Hybrid Cloud Gateway

File: _adr/89/index.adoc
Comment (machi1990) [reply _adr/89/index.adoc:57]: Sounds good. I am also open to suggestions on whether & how we could consider migrating the existing data planes to using this solution. My suggestion was that we make this change a non disruptive one to existing users / Kafkas but I am open to alternative suggestion e.g go ahead with the change all over and offer a migration guide

File: _adr/89/index.adoc
Comment (machi1990) [reply _adr/89/index.adoc:25]: took the liberty of writing down what I propose above. I'll push it for second round of review once you've the time.

**Comment by machi1990** [comment]
thanks for the review @emmanuelbernard i've pushed some updates and answer the questions. Please have another look when you get a chance to.

**Review by emmanuelbernard — APPROVED** [review]

File: _adr/89/index.adoc
Comment (pb82) [reply _adr/89/index.adoc:60]: If we change the existing certificate, would that be disruptive? From what I understand it would force client to reconnect to the brokers, but other than that, are there any side effects?

File: _adr/89/index.adoc
Comment (pb82) [reply _adr/89/index.adoc:62]: @emmanuelbernard do you still think it should be a library, even though at some point certificates will be handled by HCG as presented by Craig? We're already using a library to implement this (certmagic).

Comment (emmanuelbernard) [reply _adr/89/index.adoc:62]: I'd need to see the certmagic API level honestly. We can discuss it.

**Review by pb82 — APPROVED** [review]

File: _adr/89/index.adoc
Comment (emmanuelbernard) [reply _adr/89/index.adoc:60]: I inferred that we would need to change the Kafka URL because the Kafka broker URL is simething like `kafka-uuid-broker1.kafka.bf2.org`. We did not anticipate the need.

Comment (emmanuelbernard) [reply _adr/89/index.adoc:60]: meaning the need for separate subdomains per kafka deployments

Comment (machi1990) [reply _adr/89/index.adoc:60]: > I inferred that we would need to change the Kafka URL because the Kafka broker URL is simething like kafka-uuid-broker1.kafka.bf2.org. We did not anticipate the need.

That's the issue. 

> ah but if that's the case, why not ask for a multi named cert?

I've not investigate how this will look like but from face value it looks like we'll need to generate a certificate per kafka instance. This is not what's advertised here in the ADR, but this would solve the issue, however there are challenges that are related to this:
1. We'll like need to ask for overrrides of letsencrypt limits as generating per Kafka we are more likely to hit them often
2. In the near by future, we'll have Kafka expansion feature. Which may add some complexity to the certificate management as new brokers are scaled up / brokers are scaled down we'll need to make sure that the certificate knows of new brokers or doesn't know old brokers.

File: _adr/89/index.adoc
Comment (machi1990) [reply _adr/89/index.adoc:62]: The default certmagic API is pretty easy to work with. But we'll be building things on top of it and mostly the Vault Storage which is not offered by default. So these things that we build on top of cert magic can as well as be extracted as a library/sdk.

**Comment by machi1990** [comment]
Thanks @emmanuelbernard @shawkins @pb82 for the review.

@emmanuelbernard I do not have merge permission on the repo, can you do the honor for me? Thanks

**Pull request merged**
